quint/src/parsing/ToIrListener.ts

```diff
@@ -485,6 +514,26 @@ export class ToIrListener implements QuintListener {
     this.declarationStack.push(def, ...constructors)
   }
 
+  // Pop all the type variables in the head of a type def from the type stack
+  // E.g., for a type def like
+  //
+  // type Foo[a,b,c] = ...
+  //
+  // Return the type variables [a, b, c]
+  private popTypeDefHeadTypeVars(ctx: p.TypeDefHeadContext): QuintVarType[] {
+    return (
+      ctx._typeVars
+        .map(
+          _ =>
+            this.popType().unwrap(() =>
+              fail('internal error: type parameter parsed with no type variable')
+            ) as QuintVarType
+        )
+        // The stack stores the variables in reverse order
+        .reverse()
+    )
```

Comment:
I believe we should use `popMany` here as we are doing for most of the popping in this file. This also has a fallback that enables error recovery (improving our error reporting).
```suggestion
    return popMany(this.typeStack, ctx._typeVars().length, this.undefinedType(ctx)) as QuintVarType[]
```

Reply:
I'm a bit torn. I like reusing code, ofc. but I also find use of wrap here quite clear and it is nice to just make use of combinators when possible. But I'm more motivated by the question of correctness: this isn't actually something we should repair with a default fallback, because if there are `_typeVars` but there are not types on the stack, than our parser is broken (either ANTLR is not parsing correctly or we have broken logic in the `ToIrListener`) and we want an internal error in that case, *not* to cover over the error with a generated type. WDYT?

Reply:
> if there are `_typeVars` but there are not types on the stack, than our parser is broken

I find it hard to know things like this for sure, because ANTLR produces weird stuff when  there is a parse error. So, to be sure, I added fallbacks almost everywhere. You probably have a better understanding of the parser than I do, and this assumption seems correct, so I'm fine trusting that and moving forward. Alternatively, we could stay on the safer side and use the fallback - `undefinedType` will produce a debug line when called, so this shouldn't be invisible to us.

Reply:
Good point: if we are still getting errors, then we won't like lose track of this... 

That said, my understanding is that we cannot ever hit this method unless the generated parser successfully parses a `typeDeclaration` and any successful parse of a type declaration should either have consumed a type (put on the stack) for each type variable in the definition head. If this understanding turns out wrong, I'd like to get a hard failure and an error report rather than have (users including myself) trying to figure out why they have a parse error.

So, I'll stick with this as is since you're OK with it, and I'll be prepared to revise my understanding and replace this if we hit these assertion failures! :)